src/main/java/filter/Parser.java

```diff
@@ -17,24 +17,62 @@
 
 public final class Parser {
 
-    private Parser(ArrayList<Token> input) {
+    private final List<Token> input;
+
+    /**
+     * Determines if the parser should fail gracefully if it finds the input
+     * in a incomplete state. See {@link #check}.
+     */
+    private final boolean isInCheckMode;
+
+    private int position = 0;
+    private int sourcePosition = 0;
+
+    private Parser(List<Token> input, boolean isInCheckMode) {
         this.input = input;
+        this.isInCheckMode = isInCheckMode;
     }
+
+    /**
+     * Parses an input string, returning an abstract syntax tree.
+     * Throws a {@link ParseException} on syntax error.
+     */
     public static FilterExpression parse(String input) {
-        if (input == null || input.isEmpty()) return Qualifier.EMPTY;
-        return new Parser(new Lexer(input).lex()).parseExpression(0);
+        if (input.isEmpty()) {
+            return Qualifier.EMPTY;
+        }
+        return new Parser(new Lexer(input).lex(), false).parseExpression(0);
     }
 
-    private final ArrayList<Token> input;
-    private int position = 0;
-    private int sourcePosition = 0;
+    /**
+     * Checks an input string for errors. Returns true if the string is at
+     * least partially valid, i.e. perfectly valid, or valid but incomplete.
+     */
+    public static boolean check(String input) {
+        if (input.isEmpty()) {
+            return true;
+        }
+
+        try {
+            new Parser(new Lexer(input).lex(), true).parseExpression(0);
+            return true;
+        } catch (IncompleteInputException ignored) {
+            return true;
+        } catch (ParseException ignored) {
+            return false;
+        }
```

Comment:
Using exceptions to control 'normal' workflow is not exactly right, but may be there is no choice.